### PR TITLE
Skip the expert-parallel sentinel masking when expert parallelism is off

### PR DESCRIPTION
**Pull request by qgallouedec**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48201) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48201)

`grouped_mm_experts_forward` runs the expert-parallel sentinel dance unconditionally:

https://github.com/huggingface/transformers/blob/d56c55bf564ddb176759eb6ec199442682564916/src/transformers/integrations/moe.py#L420-L421
https://github.com/huggingface/transformers/blob/d56c55bf564ddb176759eb6ec199442682564916/src/transformers/integrations/moe.py#L437
https://github.com/huggingface/transformers/blob/d56c55bf564ddb176759eb6ec199442682564916/src/transformers/integrations/moe.py#L465

Sentinel ids are produced only by `RouterParallel._prepare_output_fn`, i.e. **only under expert parallelism**. Without it the router's `topk` cannot return an index `>= num_experts`, so `sentinel_mask` is all-`False` and the two `masked_fill_` are pure memory traffic, on tensors the size of the expert activations, twice in the forward and twice again in the backward.

This skips the whole thing when expert parallelism is off. The flag defaults to `False` where the experts module is built, and is set to `True` in `MoEParamShard.shard_param` when it shards the expert dimension, i.e. on the `grouped_gemm` plan entry that every expert-parallel plan carries. That is also where `num_experts` is rewritten to the per-rank count the sentinel mask compares against, so the two cannot disagree. A module-level plan entry would not do: llama4's expert-parallel plan has `ep_router` and `grouped_gemm` but no `moe_tp_experts`. The expert-parallel path is unchanged.

## Numbers

One `Qwen3MoeSparseMoeBlock` from `Qwen/Qwen3-30B-A3B` (128 experts, top-8, `hidden=2048`), seq 4096, bf16, fwd+bwd, single H100, `grouped_mm`:

| | fwd+bwd | peak memory |
|---|---|---|
| main | 6.18 ms | 3.46 GB |
| **this PR** | **5.22 ms (1.18×)** | 3.46 GB |

<details>
<summary>bench_sentinel.py — the script that produced the table</summary>

```python
# bench_sentinel.py — cost of the expert-parallel sentinel masking when EP is off.
import statistics
import time

import torch
from transformers import AutoConfig
from transformers.models.qwen3_moe.modeling_qwen3_moe import Qwen3MoeSparseMoeBlock

MODEL, SEQ, ITERS = "Qwen/Qwen3-30B-A3B", 4096, 30

cfg = AutoConfig.from_pretrained(MODEL)
cfg._experts_implementation = "grouped_mm"
block = Qwen3MoeSparseMoeBlock(cfg).cuda().to(torch.bfloat16)
x = torch.randn(1, SEQ, cfg.hidden_size, device="cuda", dtype=torch.bfloat16, requires_grad=True)


def run():
    torch.cuda.synchronize()
    t = time.perf_counter()
    out = block(x)
    (out[0] if isinstance(out, tuple) else out).sum().backward()
    torch.cuda.synchronize()
    return time.perf_counter() - t


for _ in range(5):
    run()
times = [run() for _ in range(ITERS)]
torch.cuda.reset_peak_memory_stats()
run()
print(f"is_expert_parallel={getattr(block.experts, 'is_expert_parallel', 'n/a')}  "
      f"{statistics.median(times) * 1e3:6.2f} ms/iter   {torch.cuda.max_memory_allocated() / 2**30:5.2f} GB")
```

</details>

End to end on 2×H100 with FSDP2 (`Qwen/Qwen3-30B-A3B`, seq 4096, per-device batch 2, LoRA r=16 on attention, `chunked_nll`):

| | step | tokens/s/GPU | peak |
|---|---|---|---|
| main | 1.927 s | 4252 | 41.3 GB |
| **this PR** | **1.798 s** | **4555 (1.07×)** | 41.3 GB |

The same 1.07× was measured before #47579 refactored this code, on an 8×H100 run of the same model, and 1.05× on `Qwen3.6-35B-A3B`.
In an 8×H100 profile of that model, `aten::masked_fill_` was 2.0% of all GPU kernel time at per-device batch 1 and **3.8% at batch 4**, entirely from these two calls.

### CODE REVIEW

**Comment by qgallouedec** [comment]
CI was right, this was broken. The gate is never True.

`install_forward` has a single call site and it doesn't pass `is_expert_parallel`, so it always takes the `False` default. The masking was skipped unconditionally, including under EP, which is exactly when it's needed.

Repro, 2 GPUs, `torchrun --nproc_per_node=2 repro.py`:

```python
# torchrun --nproc_per_node=2 repro.py
import os
import tempfile

import torch
import torch.distributed as dist

from transformers import AutoModelForCausalLM, Qwen3MoeConfig
from transformers.distributed import DistributedConfig

dist.init_process_group(backend="nccl")
rank, world = dist.get_rank(), dist.get_world_size()
torch.cuda.set_device(rank)

path = os.path.join(tempfile.gettempdir(), "tiny_qwen3_moe")
if rank == 0:
    config = Qwen3MoeConfig(
        vocab_size=256, hidden_size=128, intermediate_size=256, moe_intermediate_size=64,
        num_hidden_layers=2, num_attention_heads=4, num_key_value_heads=2,
        num_experts=8, num_experts_per_tok=2, decoder_sparse_step=1, tie_word_embeddings=False,
    )
    torch.manual_seed(0)
    AutoModelForCausalLM.from_config(config, dtype=torch.bfloat16).save_pretrained(
        path, save_original_format=True
    )
dist.barrier()

model_ep = AutoModelForCausalLM.from_pretrained(
    path, dtype=torch.bfloat16,
    distributed_config=DistributedConfig(tp_size=world, enable_expert_parallel=True),
)
dist.barrier()
model_ref = AutoModelForCausalLM.from_pretrained(path, dtype=torch.bfloat16).to(model_ep.device)

torch.manual_seed(0)
input_ids = torch.randint(0, 256, (2, 64), device=model_ep.device)
with torch.no_grad():
    logits_ref = model_ref(input_ids).logits
    logits_ep = model_ep(input_ids).logits

if rank == 0:
    experts = model_ep.model.layers[0].mlp.experts
    print(f"experts implementation: {model_ep.config._experts_implementation}")
    print(f"is_expert_parallel:     {getattr(experts, 'is_expert_parallel', '<absent>')}")
    print(f"NaN in EP logits:       {bool(torch.isnan(logits_ep).any())}")
    print(f"max abs diff vs non-EP: {(logits_ref.float() - logits_ep.float()).abs().max().item()}")

dist.destroy_process_group()
```

On this branch:

```
experts implementation: grouped_mm
is_expert_parallel:     False
NaN in EP logits:       True
max abs diff vs non-EP: nan
```

With the flag actually plumbed through:

```
experts implementation: grouped_mm
is_expert_parallel:     True
NaN in EP logits:       False
max abs diff vs non-EP: 0.0
```

On main (no gate, masking unconditional) the same script gives no NaN and 0.0.

Fix is to pass the flag from `apply_tensor_parallelism`, where `model.config.distributed_config.enable_expert_parallel` is already set, plus an `is_expert_parallel=False` kwarg on the base `TensorParallelLayer.install_forward` and the two overrides that share that call site. 6 lines.

Worth flagging separately: the EP mixin tests pass on CPU even without the fix, so the CPU suite doesn't cover this path. Only the GPU run above surfaces it.

File: src/transformers/integrations/moe.py [hunk 1]
```diff
@@ -434,7 +436,8 @@ def grouped_mm_experts_forward(
         selected_biases = self.up_proj_bias[expert_ids_g] if self.has_bias else None
 
     # Pre-mask (bwd path).
-    selected_hidden_states_g.masked_fill_(sentinel_mask, 0.0)
+    if sentinel_mask is not None:
```
Comment (IlyasMoutawwakil) [inline src/transformers/integrations/moe.py]: can continue to check on "if self.is_expert_parallel:"

**Comment by IlyasMoutawwakil** [comment]
thanks for the optimization ! i remember measuring it for inference only and seeing only noise 🥲

**Comment by IlyasMoutawwakil** [comment]
not sure but i think we can probably know if EP is enabled by checking for DTensor + sharding dim (0) of gate_up/down proj

**Comment by IlyasMoutawwakil** [comment]
btw there are other implementations that do the same, in finegrained-fp8 and deepgemm

File: src/transformers/integrations/moe.py
Comment (IlyasMoutawwakil) [inline src/transformers/integrations/moe.py]: here as well 🙏

**Comment by qgallouedec** [comment]
@IlyasMoutawwakil I tried the `DTensor+Shard(0)` derivation and it can't work: `_use_local_dtensor_params` wraps the experts forward under EP, so inside the forward the weights are already plain local tensors and the check reads False when EP is on.

(to verify: an in-forward probe shows `type(gate_up_proj) == Tensor` under EP, and the skipped clamp makes `batched_mm` crash with an OOB gather.)

Kept the explicit attribute instead, applied your other two points: the batched_mm clamp is now gated too, and the same skip is applied in finegrained-fp8 and deepgemm.

**Comment by qgallouedec** [comment]
CI was red on nine kernel tests, all the same cause: the tests build their experts stand-in as a `SimpleNamespace`, so it never goes through `use_experts_implementation.__init__`
Fixed in eb9dafc63f, full tests/kernels/ run should be green now.

File: src/transformers/distributed/tensor_parallel.py [hunk 2]
```diff
@@ -816,7 +817,9 @@ def apply_tensor_parallelism(model, tp_mesh):
                 # MLA needs to know the qk_rope_head_dim to split the projection output into KV and RoPE parts.
                 # TODO: Store qk_rope_head_dim on MLA projection modules when the models initialize them.
                 module.config = model.config.get_text_config()
-            ALL_PARALLEL_STYLES[style_name].install_forward(module, tp_mesh)
+            ALL_PARALLEL_STYLES[style_name].install_forward(
+                module, tp_mesh, is_expert_parallel=model.config.distributed_config.enable_expert_parallel
+            )
```
Comment (VI-Arthur) [inline src/transformers/distributed/tensor_parallel.py]: not a fan of adding a new parameter here just for this!

File: src/transformers/distributed/tensor_parallel.py [hunk 3]
```diff
@@ -606,6 +606,7 @@ def transform_inputs_pre_forward(self, module, args, kwargs, mesh, *, is_expert_
 
     def install_forward(self, module, mesh, *, is_expert_parallel=False):
         """Install the transforms but pass `is_expert_parallel` in the forward call."""
+        module.is_expert_parallel = is_expert_parallel
```
Comment (VI-Arthur) [inline src/transformers/distributed/tensor_parallel.py]: you have access to module.config afaik -> can be simpler to just take it to extract parallel flag?

**Review by VI-Arthur — APPROVED** [review]
LGTM, much needed indeed cc @IlyasMoutawwakil

File: src/transformers/distributed/tensor_parallel.py (hunk 3 above)
Comment (qgallouedec) [reply src/transformers/distributed/tensor_parallel.py]: Fair, dropping the parameter.

`module.config` doesn't quite work: it's only set a few lines above for `mla_kv_a_proj`, so it isn't there for the other styles.
But the parameter was only ever used to set `module.is_expert_parallel`, so the caller sets that directly now and `install_forward` goes back to `(self, module, mesh)` on all four styles

File: src/transformers/distributed/tensor_parallel.py (hunk 2 above)
Comment (qgallouedec) [reply src/transformers/distributed/tensor_parallel.py]: see https://github.com/huggingface/transformers/pull/48201#discussion_r3960109545

**Comment by qgallouedec** [comment]
@VI-Arthur needs another look, the version you approved had a bug.

Passing the flag through `install_forward` also reached

https://github.com/huggingface/transformers/blob/2ec1d59ce66a3be6e7c41931a5d44fc0c84dcd75/src/transformers/distributed/tensor_parallel.py#L601-L602

where a true value skips the `_AllReduceBackward` on `top_k_weights`: router grads would stop being summed under EP. That path is back to main.

Flag is now set in `MoEParamShard.shard_param` under `shards_expert_dim`, next to the `num_experts` rewrite the mask compares against. A module-level entry doesn't work, llama4's ep_plan has no `moe_tp_experts`. That's the bit I'd like your eyes on.

Also dropped the parameter from the three overrides that ignored it, per your comment. `tensor_parallel.py` is now main + 2 lines.

**Review by IlyasMoutawwakil — APPROVED** [review]
lgtm ! just make sure the new flag will stay relevant in the a2a PR where there are no sentinels iiuc

File: src/transformers/distributed/tensor_parallel.py [hunk 4]
```diff
@@ -541,6 +541,8 @@ def shard_param(self, module, param, mesh):
                     f"{expert_parallel_size} expert-parallel ranks."
                 )
             module.num_experts = global_num_experts // expert_parallel_size
+            # The experts forward masks sentinel rows only when its experts are actually split.
+            module.is_expert_parallel = True
```
Comment (IlyasMoutawwakil) [inline src/transformers/distributed/tensor_parallel.py]: ah yes just one thing, do you need this public ? it would be bc if we remove it/replace it later and the info is in the dist config already for users

**Review by IlyasMoutawwakil — APPROVED** [review]

File: src/transformers/distributed/tensor_parallel.py (hunk 4 above)
Comment (qgallouedec) [reply src/transformers/distributed/tensor_parallel.py]: Fair. The name doesn't have to be public. Mostly your call on this, just tell me

Comment (IlyasMoutawwakil) [reply src/transformers/distributed/tensor_parallel.py]: yes let's keep it internal

Comment (qgallouedec) [reply src/transformers/distributed/tensor_parallel.py]: Done in ca7eeca34a, it is `_is_expert_parallel` now.
